### PR TITLE
Optimize initial dashboard load - load dashboard metadata in parallel with the query

### PR DESCRIPTION
**Pull request by ranquild**
Follow up for https://github.com/metabase/metabase/pull/42901

Makes `query` and `query_metadata` start in parallel, but wait until metadata is loaded because the UI won't work until then.

Screenshot 2024-05-24 at 11 16 37

### CODE REVIEW

File: frontend/src/metabase/dashboard/actions/data-fetching.js [hunk 1]
```diff
@@ -376,17 +380,16 @@ export const fetchDashboardCardData =
...
   };
 
-export const fetchDashboardCardMetadata = createThunkAction(
+export const fetchDashboardCardMetadata = createAsyncThunk(
   FETCH_DASHBOARD_CARD_METADATA,
-  () => async (dispatch, getState) => {
-    const allDashCards = getDashboardComplete(getState()).dashcards;
-    const selectedTabId = getSelectedTabId(getState());
-
-    const cards = allDashCards.filter(
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/data-fetching.js:385]: It's much more efficient to load all metadata at once due to how MBQL lib works

File: frontend/src/metabase/dashboard/actions/data-fetching-typed.ts [hunk 2]
```diff
@@ -124,10 +123,6 @@ export const fetchDashboard = createAsyncThunk(
 
       fetchDashboardCancellation = null;
 
-      if (dashboardType === "normal" || dashboardType === "transient") {
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/data-fetching-typed.ts:127]: We now don't block `query` endpoints until metadata is loaded. The same check is moved to `fetchDashboardCardMetadata`

File: frontend/src/metabase/dashboard/hoc/DashboardData.jsx [hunk 3]
```diff
@@ -110,7 +114,6 @@ export const DashboardData = ComposedComponent =>
 
         if (!_.isEqual(nextProps.selectedTabId, this.props.selectedTabId)) {
           this.props.fetchDashboardCardData();
-          this.props.fetchDashboardCardMetadata();
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/hoc/DashboardData.jsx:113]: We don't need to load metadata when the tab changes now

File: frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx [hunk 4]
```diff
@@ -321,7 +323,6 @@ function DashboardInner(props: DashboardProps) {
     }
     if (previousTabId !== selectedTabId) {
       fetchDashboardCardData();
-      fetchDashboardCardMetadata();
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx:324]: We don't need to load metadata when the tab changes now

File: frontend/src/metabase/public/containers/PublicDashboard/PublicDashboard.tsx [hunk 5]
```diff
@@ -158,7 +160,6 @@ class PublicDashboardInner extends Component<PublicDashboardProps> {
 
     if (!_.isEqual(prevProps.selectedTabId, this.props.selectedTabId)) {
       this.props.fetchDashboardCardData();
-      this.props.fetchDashboardCardMetadata();
```
Comment (ranquild) [inline frontend/src/metabase/public/containers/PublicDashboard/PublicDashboard.tsx:161]: We don't need to load metadata when the tab changes now

**Review by uladzimirdev — APPROVED** [review]
🤞

**Pull request merged**